File: ros1bookcode-master-rw/chapt10/chapt10_ws/src/learn_message_filter_cpp/src/timesync_test.cpp
#include <ros/ros.h>
#include <message_filters/subscriber.h>
#include <message_filters/synchronizer.h>
#include <message_filters/sync_policies/approximate_time.h>
#include <sensor_msgs/Imu.h>
#include <nav_msgs/Odometry.h>

using Imu = sensor_msgs::Imu;
using Odometry = nav_msgs::Odometry;
using MySyncPolicy = message_filters::sync_policies::ApproximateTime<Imu, Odometry>;

class TimeSyncTestNode {
public:
  TimeSyncTestNode() : imu_sub_(nh_, "imu", 10), odom_sub_(nh_, "odom", 10) {
    raw_imu_sub_ = nh_.subscribe("imu", 10, &TimeSyncTestNode::imuCallback, this);
    raw_odom_sub_ = nh_.subscribe("odom", 10, &TimeSyncTestNode::odomCallback, this);
    synchronizer_.reset(new message_filters::Synchronizer<MySyncPolicy>(MySyncPolicy(10), imu_sub_, odom_sub_));
    synchronizer_->registerCallback(boost::bind(&TimeSyncTestNode::resultCallback, this, _1, _2));
  }
private:
  void imuCallback(const Imu::ConstPtr& imu_msg) { ROS_INFO("imu(%u,%u)", imu_msg->header.stamp.sec, imu_msg->header.stamp.nsec); }
  void odomCallback(const Odometry::ConstPtr& odom_msg) { ROS_INFO("odom(%u,%u)", odom_msg->header.stamp.sec, odom_msg->header.stamp.nsec); }
  void resultCallback(const Imu::ConstPtr& imu_msg, const Odometry::ConstPtr& odom_msg) {
    ROS_INFO("imu(%u,%u),odom(%u,%u)", imu_msg->header.stamp.sec, imu_msg->header.stamp.nsec, odom_msg->header.stamp.sec, odom_msg->header.stamp.nsec);
  }
  ros::NodeHandle nh_;
  ros::Subscriber raw_imu_sub_, raw_odom_sub_;
  message_filters::Subscriber<Imu> imu_sub_;
  message_filters::Subscriber<Odometry> odom_sub_;
  std::shared_ptr<message_filters::Synchronizer<MySyncPolicy>> synchronizer_;
};

int main(int argc, char** argv) {
  ros::init(argc, argv, "sync_node");
  TimeSyncTestNode node;
  ros::spin();
  return 0;
}

File: ros1bookcode-master-rw/chapt10/chapt10_ws/src/learn_message_filter_py/learn_message_filter_py/timesync_test.py
#!/usr/bin/env python3
import rospy
from sensor_msgs.msg import Imu
from nav_msgs.msg import Odometry
from message_filters import Subscriber, ApproximateTimeSynchronizer

class TimeSyncTestNode:
    def __init__(self):
        self.imu_sub = Subscriber('imu', Imu)
        self.odom_sub = Subscriber('odom', Odometry)
        self.synchronizer = ApproximateTimeSynchronizer([self.imu_sub, self.odom_sub], 10, 0.01)
        self.synchronizer.registerCallback(self.result_callback)
        rospy.Subscriber('imu', Imu, self.imu_callback, queue_size=10)
        rospy.Subscriber('odom', Odometry, self.odom_callback, queue_size=10)
    def imu_callback(self, msg):
        rospy.loginfo(f'imu({msg.header.stamp.secs},{msg.header.stamp.nsecs})')
    def odom_callback(self, msg):
        rospy.loginfo(f'odom({msg.header.stamp.secs},{msg.header.stamp.nsecs})')
    def result_callback(self, imu_msg, odom_msg):
        rospy.loginfo(f'imu({imu_msg.header.stamp.secs},{imu_msg.header.stamp.nsecs}),odom({odom_msg.header.stamp.secs},{odom_msg.header.stamp.nsecs})')

def main():
    rospy.init_node('sync_node')
    TimeSyncTestNode()
    rospy.spin()
if __name__ == '__main__':
    main()
